### ISSUE TITLE
Samples seem to get lost across event loop runs

### ISSUE THREAD

**Issue by thlorenz** [thlorenz/sample-heap#1]
When running an example with a `setImmediate` call in it (starting sampling and collecting samples on the next uv loop run) we seem to be loosing all samples.

To repro, checkout [the better-tree branch](https://github.com/thlorenz/sample-heap-allocations/tree/better-tree) (a heavily WIP) and run the `./example/async.js` similar to:

``` js
console.log('sync')
sp.startSampling()
allocate.flatArray()
printAllocs()

console.log('async')
sp.startSampling()
allocate.flatArray()
setImmediate(printAllocs)
function printAllocs() {
  const infos = sp.collectAllocations()
  inspect(infos)
}
```

You'll get this output:

```
sync
{ id: '0:0:0',
  script_id: 0,
  script_name: '',
  name: '(root)',
  line_number: 0,
  column_number: 0,
  start_position: 0,
  allocations: [],
  child_ids: [ '34:10:10' ],
  children: 
   [ { id: '34:10:10',
       script_id: 34,
       script_name: 'node.js',
       name: '',
       line_number: 10,
       column_number: 10,
       start_position: 308,
       allocations: [],
       child_ids: [Object],
       children: [Object] } ] }
async
{ id: '0:0:0',
  script_id: 0,
  script_name: '',
  name: '(COMPILER)',
  line_number: 0,
  column_number: 0,
  start_position: 0,
  allocations: [],
  child_ids: [],
  children: [] }
```

I looked at all the sample memory allocations tests in the v8 code base and all run in sync, i.e. never leave/re-enter JavaScript.
So at this point I'm not sure if this behavior is intended, a bug and/or if there is a workaround.

Obviously in the case of Node.js I'd want to collect samples and have them persist across those JS re-entries as I might be handling async requests while I'm collecting them.

Will dive deeper into the sampling code to find an answer, but I hoped that @ofrobots could help me here as you know a lot more about this than I do at this point. I'd greatly appreciate it :)

**Comment by thlorenz** [comment]
Just realizing that I must have started sampling on one loop run and collected on another when I created the first demo app [demoed in this gif](https://github.com/thlorenz/sample-heap-allocations/blob/master/assets/sampling-memory.gif).
So at this point not sure what changed fro then to now. I did _hopefully_ improve the way I collect the data into a tree, but could be that I introduced a mistake there, so will investigate on my end as well.
